Signal/src/call/PeerConnectionClient.swift

```diff
@@ -205,111 +233,161 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
     }
 
     public func setAudioEnabled(enabled: Bool) {
-        guard let audioTrack = self.audioTrack else {
-            let action = enabled ? "enable" : "disable"
-            Logger.error("\(TAG) trying to \(action) audioTrack which doesn't exist.")
-            return
-        }
+        AssertIsOnMainThread()
+
+        PeerConnectionClient.signalingQueue.async {
+            guard let audioTrack = self.audioTrack else {
+                let action = enabled ? "enable" : "disable"
+                Logger.error("\(self.TAG) trying to \(action) audioTrack which doesn't exist.")
+                return
+            }
 
-        audioTrack.isEnabled = enabled
+            audioTrack.isEnabled = enabled
+        }
     }
 
     // MARK: - Session negotiation
 
-    var defaultOfferConstraints: RTCMediaConstraints {
+    private var defaultOfferConstraints: RTCMediaConstraints {
         let mandatoryConstraints = [
             "OfferToReceiveAudio": "true",
             "OfferToReceiveVideo": "true"
         ]
         return RTCMediaConstraints(mandatoryConstraints:mandatoryConstraints, optionalConstraints:nil)
     }
 
-    func createOffer() -> Promise<HardenedRTCSessionDescription> {
-        return Promise { fulfill, reject in
-            peerConnection.offer(for: self.defaultOfferConstraints, completionHandler: { (sdp: RTCSessionDescription?, error: Error?) in
-                guard error == nil else {
-                    reject(error!)
-                    return
-                }
-
-                guard let sessionDescription = sdp else {
-                    Logger.error("\(self.TAG) No session description was obtained, even though there was no error reported.")
-                    let error = OWSErrorMakeUnableToProcessServerResponseError()
-                    reject(error)
-                    return
-                }
+    public func createOffer() -> Promise<HardenedRTCSessionDescription> {
+        AssertIsOnMainThread()
 
-                fulfill(HardenedRTCSessionDescription(rtcSessionDescription: sessionDescription))
-            })
+        return Promise { fulfill, reject in
+            AssertIsOnMainThread()
```

Comment:
Promise-related code can be hard to read; the concurrency in particular isn't always self-evident.  I've been prolific with asserts to make this more clear.